bee-message/src/parents.rs

```diff
@@ -0,0 +1,148 @@
+// Copyright 2021 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+//! Module that provides types and syntactic validations of message parents.
+
+use crate::{MessageId, MessageUnpackError, ValidationError};
+
+use bee_ord::is_unique_sorted;
+use bee_packable::{coerce::*, PackError, Packable, Packer, UnpackError, Unpacker};
+
+use alloc::vec::Vec;
+use core::{
+    convert::{Infallible, TryFrom, TryInto},
+    ops::RangeInclusive,
+};
+
+/// The range representing the valid number of parents.
+pub const MESSAGE_PARENTS_RANGE: RangeInclusive<usize> = 1..=8;
+
+/// Minimum number of strong parents for a valid message.
+pub const MESSAGE_MIN_STRONG_PARENTS: usize = 1;
+
+/// An individual message parent, which can be categorized as "strong" or "weak".
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+#[cfg_attr(
+    feature = "serde1",
+    derive(serde::Serialize, serde::Deserialize),
+    serde(tag = "type", content = "data")
+)]
+#[repr(u8)]
+pub enum ParentsKind {
+    /// Message parents in which the past cone is "Liked".
+    Strong = 0,
+    /// Message parents in which the past cone is "Disliked", but the parents themselves are "Liked".
+    Weak = 1,
+    /// Message parents that are "Liked".
+    Disliked = 2,
+    /// Message parents that are "Disliked".
+    Liked = 3,
+}
+
+impl TryFrom<u8> for ParentsKind {
+    type Error = ValidationError;
+
+    fn try_from(value: u8) -> Result<Self, Self::Error> {
+        match value {
+            0 => Ok(Self::Strong),
+            1 => Ok(Self::Weak),
+            2 => Ok(Self::Disliked),
+            3 => Ok(Self::Liked),
+            _ => Err(ValidationError::InvalidParentsKind(value)),
+        }
+    }
+}
+
+/// A block of message parent IDs, all of the same [`ParentsKind`].
+///
+/// [`ParentsBlock`]s must:
+/// * Be of a valid [`ParentsKind`].
+/// * Contain a valid count of parents (1..=8).
+/// * IDs must be unique and lexicographically sorted in their serialized forms.
+#[derive(Clone, Debug, PartialEq, Eq)]
+#[cfg_attr(feature = "serde1", derive(serde::Serialize, serde::Deserialize))]
+pub struct ParentsBlock {
+    kind: ParentsKind,
+    references: Vec<MessageId>,
```

Comment:
This could be a `VecPrefix<MessageId, u8>`  but `u8` doesn't implement `Bounded` can we overcome this restriction @Adam-Gleave ?

Reply:
Can this not be a `VecPrefix<MessageId, BoundedU8<1, 4>>`?